File: 2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *newHead = new ListNode(-1);
        ListNode *tmpNode = newHead;
        int last = 0;
        while(l1 || l2){
            int t = last;
            if(l1){
                t += l1->val;
                l1 = l1->next;
            }
            if(l2){
                t += l2->val;
                l2 = l2->next;
            }
            tmpNode->next = new ListNode(t%10);
            tmpNode = tmpNode->next;
            last = t/10;
        }
        if(last){
            tmpNode->next = new ListNode(last);
        }
        return newHead->next;
    }
};

File: go-solutions/2.go
package main

//Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var newHead *ListNode
	tmp := newHead
	last := 0
	for {
		if l1 == nil && l2 == nil {
			break
		}
		val := last
		if l1 != nil {
			val += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			val += l2.Val
			l2 = l2.Next
		}
		tmp.Next = new * ListNode
		tmp.Next.Val = val % 10

		last = val / 10
		tmp = tmp.Next
	}
	if last != 0 {
		tmp.Next = new * ListNode
		tmp.Next.Val = last % 10
	}
	return newHead.Next
}
